### ISSUE TITLE
Firemonkey: source code

### ISSUE THREAD

**Issue by atomGit** [erosman/support#234]
the MPL states...

> such Covered Software must also be made available in Source Code Form

i realize the source is available in the .xpi - is it published elsewhere? i couldn't find it at github, gitlab or codeberg

**Comment by erosman** [comment]
> i realize the source is available in the .xpi - 

Correct

> is it published elsewhere? i couldn't find it at github, gitlab or codeberg

No. I don't use them to develop
Previously: #97 #110

**Comment by atomGit** [comment]
for me this is rather trivial, but not publishing the source doesn't conflict with the MPL?

aside from that, publishing the source allows one to view commits

i understand and respect that you have your own workflow, but the problem i see is that SO MANY of the add-ons at AMO are what i would call malware in that they do data collection and various other nefarious stuff, or a developer will sell-out to a 3rd party or otherwise find a non-ethical way to monetize their add-on, thus why it's really nice to view the code commits and verify, separate from what the developer says, that nothing funny is going on

while this is of coarse doable by extracting the XPI and comparing versions, that is a time consuming chore that isn't practical

so i get that you have a preferred workflow and indeed appear to be an ethical person, however i respectfully ask that you (re)consider publishing the source for the sake of our your users sanity :)

if you use [codeberg](https://codeberg.org/), they have no 3rd party dependencies

**Comment by erosman** [comment]
The code is not minified and easy to read. Extracting a xpi is not very cumbersome for those who are programmers and already familiar with process.

Non-programmers don't really check the code any way.

I regularly change the code and reflecting each change on a remote site is not feasible.
I also develop, use and road test at the same time which is not possible when developing via internet sites.

**Comment by rezad1393** [comment]
wow this is such a bad thing in today's software world.
I got to this addon from ghacks site and I was so exited for an addon that uses official firefox api.
but using a binary with not even a non-git source code is no longer appealing to me.

it is not just the malware concern, but also code continuity and code sharing.
after using firefox for more that 15 years I cant be bothered to use addons that there are no hope of forking them if the main developer just disappears or gets tired of it.

I hope this addon owner also uses the FOSS model and let others also share their knowledge and time for development and betterment of this addon.

as the other guy said it is not just about extracting xpi. it is about having an version controlled accessible shared development that assures users that addon would not be easily abandoned.

**Comment by erosman** [comment]
> ..... version controlled accessible shared development ....

@rezad1393 There are those who develop using online tools such as Github etc. This model is best suited for shared code development where few people contribute to the code and changes do not occur too frequently.

For a single developer, who often rewrites code 10-20 times a day and makes changes, it is not possible to development via an online method.

Forking github has its pro & cons. While it helps in case a developer drops a project, in reality it has become the forte of plagiarists. As an example, there has been 100s of forks of _uBlock Origin_ that have been uploaded to AMO by people who are not programmers (copy-pasters) and/or people who hide malicious code within good code.

There are also open-source extensions that have their code published on Github which is good but the extension code itself is minified and requires additional scripts and processes to match the code of the extension to the code published on Github. 

FOSS require the software to be free and open and inspectable, and FireMonkey meets those requirements.

**Comment by rezad1393** [comment]
thank you for the long and good answer. English is not my  main language so maybe I was not that clear.

when I said FOSS model I meant in spirit not in strict intreperatation. I meant allowing other to be involved so that humanity as a whole benefits and we dont invent the wheel too many times. yes I mean that sincerely, and I believe that software/algorithm  duplicates waste a lot of man/hours (well man/years to be honest) of humanity potential.

also I understand a lot of point you make but I cant agree with most of them.

for example (a negative example) amd graphic driver for linux are supposed to be opensource (and kinda FOSS) but amd doesnt actually share the code history. they only upload versions to github.
that is a example of a bad FOSS project that wastes its own potential if it was shared with public to participate it code review in every commit.
but I understand that they have to do it because of corporate issues that may pop up if they share every commit (or at least I think that is one of the reasons) 

about the single developer part:
even if nobody else ( and I mean literally nobody else) helped you with this it is still beneficiary to put the code here so that other benefit from it. ease of access for new programmers is very important and if they can search github (or any other code hosting for that matter ) and find your code they can learn a lot from it.
also you dont need to push all your changes to github (As far as I am familiar with git).
committing to git repo on github every few days the aggregates changes is not against the law :) .

also you cant stop those who want to (as you said) plagiarize it. even ublock couldnt .
not uploading the code to github doesnt stop that.
it actually encourages it  because if you had a git history on github of your code changes you could show that you are the main (not-plagiarist) version and in extreme cases you can submit removal request to github or Mozilla for those that copy your code when they dont follow the copy-right part of FOSS (the parts about showing who the real author of original code is and so on) 

but in the end it is your project and you are to do as you think is best.
these are just suggestions that I just gathered from my experience with software as a end user and seeing the state of opensource software in relation to updates and codechanges and so on.

have a good day.

**Comment by erosman** [comment]
Thank you for the feedback :+1
Let's see what future holds.

**Comment by SirWindfield** [comment]
> I regularly change the code[...].

Yes, but that doesn't prevent you from using a version control system (VCS) like `git` or `svn`. Note that `git` is not GitHub. `git` is used locally to create a repository, add files to it, commit changes and do other stuff related to development. As soon as you finished implementing a feature or fixing a bug you can simply commit the changes to the repository. No one tells you how often you have to commit. But if you'd be willing to publish a new version of the addon then you'd probably should create a commit at that point. I don't see why this would break your workflow to be honest.

Some kind of VCS should be inplace to prevent accidental breakage during programming anyway. Be it copy-pasting folders (please don't), or a traditional VCS.

> [...] and reflecting each change on a remote site is not feasible.

Well, if you use a VCS like `git` or `svn`, that is free-of-charge. `git push` does it for you. Done.

> I also develop, use and road test at the same time which is not possible when developing via internet sites.

These seems more to me that you got the idea wrong. You don't develop on any internet site. GitHub, Gitlab, BitBucket or merely mirrors of your local repository. What you do locally is up to you. You just mirror those changes and files onto a server.

I don't mean this in to be offensive in any way, I just wanted to clear up some stuff as I had the feeling that you didn't understand _some_ concepts of VCS and remote servers. Millions of developers use internet services like GitHub to publish their source code, work on their projects locally and test them there as well. And it all works for them. If you're simply not familiar with VCS in general, I'd highly recommend to get into it. And there is nothing better thn a project of yours, even if you don't upload your changes onto a remote server.

If you simply don't want to publish the source code "officially", that's a totally different story and that decision is up to you for sure. It's your project, your intellectual property and your decision on how to handle it. And you shouldn't feel forced to upload your code just because some people request it. In that case, maybe a pinned issue stating that you don't plan to publish the code for the foreseeable future might be a good idea. There have been a lot of people asking the question by now. 

From my standpoint: If a browser extension tampers with literally every site that I interact with, I want to double check that it doesn't do any malicious things or collects usage data. I'd rather clone a repository, take a look locally, and build the extension myself (or load it from a folder) than having good faith and download it from the add-on store. I can do that with the steps provided by you above, but having the source code is just another step of transparency (that I and certainly others would appreciate :) ).

**Comment by erosman** [comment]
Thank you @SirWindfield for the feedback.

> Well, if you use a VCS like git or svn, that is free-of-charge. git push does it for you. 

I use a plain editor like Notepad++, to edit & run unpacked addon live on my Nightly (Win7).  In other words, with only an editor & Nightly. How does that set-up run VCS without installing new applications?

Please consider that ...
- Programmers, shouldn't have any problem extracting the XPI and inspecting the code
- Non-programmers would find it difficult to inspect the code even if it was on Github

Furthermore ...
- Tampermonkey with the largest user-base in user-script mangers on Firefox (421,000+ users) & Chrome (10,000,000+ users), not only doesn't `git` its code, but is Proprietary License and minified
- Stylish with the largest user-base in user-style mangers on Firefox (74,000+ users) & Chrome (2,000,000+ users), not only doesn't `git` its code, but is Proprietary License and minified
- Violentmonkey on Firefox (43,000+ users) & Chrome (600,000+ users) code is minified and a compile script is needed to compare the addon code with the published github code
- Greasemonkey on Firefox (283,000+ users) code is clear and on github but it is worked on by more than one person

AFA Tampermonkey & Stylish, nearly 12.5 million users don't mind not seeing the code. FireMonkey only has 440 users ATM. :wink: 

I am happy to answer any queries regarding the code and will consider all code improvement suggestions from programmers and interface suggestions from users.

_PS. FireMonkey is one of my 14 addons on AMO. My two recommended addons with combined 38+k users have not been subjected to `git` code publication requests._

**Comment by atomGit** [comment]
> I use a plain editor l ...

which you can continue to do - git does not require any new software other than git which is free code and transparent to the developer except when you use it commit/push to the repo in which case it makes it a lot easier to do accurately and efficiently

> Programmers, shouldn't have any problem extracting the XPI and inspecting the code

and comparing it to what? should every coder maintain their own git repo instead of you having 1 where they can easily read the commits?

lastly, i would not recommend github since it's now owned by fucking microsoft - instead there's gitlab, or better yet, [codeberg](https://codeberg.org/)

**Comment by erosman** [comment]
> git does not require any new software

That means installing git, doesn't it? 
Therefore, I need to install additional software i.e. git, to accomplish this task.
It also uses CLI which is not my forte.

>  comparing it to what?

Notepad++ can compare. There are other compare tools available.
I do it all the time to compare a version to another.

> ... github since it's now owned by ... microsoft

:man_facepalming:

**Comment by SirWindfield** [comment]
> That means installing git, doesn't it?
Therefore, I need to install additional software i.e. git, to accomplish this task.
It also uses CLI which is not my forte.

> Please consider that ... Programmers, shouldn't have any problem extracting the XPI and inspecting the code [...].

That's true. You'd have to install `git`. However, it's super easy to install on Unix/Linux. `sudo apt install git` on Ubuntu for example. Given that you're not used to a terminal, you're probably using macOS or Windows. On macOS it's preinstalled and on Windows it's a simple installable: https://git-scm.com/downloads. 
There are also good (and open-source) GUI clients for `git` that make it really easy to get started alongside a good tutorial on the internet. I'll link them below if you're curious :)

> Therefore, I need to install additional software i.e. [...].

Note that it's almost always a given that a developer (nowadays) has `git` installed. It's the goto tool for VCS.

I'm still glad that you took your time and read through the comment @erosman :)

GitHub Desktop: [open-source](https://github.com/desktop/desktop) https://desktop.github.com/
GitKraken: https://www.gitkraken.com/

**Comment by erosman** [comment]
I am on Windows.
I will check those out.

**Comment by rezad1393** [comment]
> I am on Windows.
> I will check those out.

at the risk of not being related to topic, why are you on windows 7?
is is hardware limits? as in, windows 10 doesnt install on you hardware?
if so may I recommend linux?
I tried linux 16 years ago and haven't looked back.
unless you need windows for some specific application that has no linux alternative (maybe and engineering application) I highly (very highly) recommend linux.

also I think you get the vcs concept wrong.
it is about tracking changes in a project.
say you add some feature and remove others at a specific date. then later you want to re-add removed feature. the git (or any version control system) would allow you to do that and check for you if you need to change a lot of files or not.
again: git is not just for collaboration between more than one person on a project.
I am not a programmer but every starter programming guide that I saw in the last 10 years start with learning a version control system (usually git because it is foss ,created by linux creator, linus trovalds) .
it actually helps starter and intermediate programmers more than very professional one because it shows your own history of change and you can learn a lot from that.

I really really hope I got you wrong and you didnt mean that you just use a text editor with no vcs used.

**Comment by erosman** [comment]
> at the risk of not being related to topic, why are you on windows 7?
> is is hardware limits? as in, windows 10 doesnt install on you hardware?

Win10 will be a bit heavy for my i5 with 4gb RAM :disappointed:

> if so may I recommend linux?
> I tried linux 16 years ago and haven't looked back, and I am NOT a programmer just a user.
> unless you need windows for some specific application that has no linux alternative (maybe a engineering application) I highly (very highly) recommend linux.

TBH, I tried Linux (dual boot) 20+ years ago (Slackware, Gentoo, Debian) and it was not easy for me and the kernel panic got me frustrated.

However, I have been thinking about Linux for a few years now. I have been looking for a stable (no kernel panic) distro but everybody has a different opinion on the most suitable Linux for my use.

_PS. I will reopen this topic for more feedback._

**Comment by rezad1393** [comment]
> 
> Win10 will be a bit heavy for my i5 with 4gb RAM disappointed
> 
> > if so may I recommend linux?
> > I tried linux 16 years ago and haven't looked back, and I am NOT a programmer just a user.
> > unless you need windows for some specific application that has no linux alternative (maybe a engineering application) I highly (very highly) recommend linux.
> 
> TBH, I tried Linux (dual boot) 20+ years ago (Slackware, Gentoo, Debian) and it was not easy for me and the kernel panic got me frustrated.
> 
> However, I have been thinking about Linux for a few years now. I have been looking for a stable (no kernel panic) distro but everybody has a different opinion on the most suitable Linux for my use.
> 
> _PS. I will reopen this topic for more feedback._

windows 10..........
whenever I have to touch a computer with a windows 10 on it I shudder and disinfect my hands and take a shower. 
:)


I use archlinux but for a beginner it may be too much at first.
about linux I recommend ubuntu for beginners and maybe fedora for intermediate.( two very different philosophies for their creators).
ubuntu is tried and tested and the first linux distro that got massive adaptation (based on the great stable debian distro)

linux works good with low ram and it is (in my opinion) way more simple than windows to use and learn.

if you are a person that (based on your post history for editor choice) a kinda minimalist, then linux is THE undisputed king of that way.

if you tested linux 20 years ago , there is no comparison  with linux today.
so I say go for it and maybe later we can convert you to git too ;) 
based on your spec (4GB ram) then the laptop (or pc) is not a weird config so the chance that linux doesnt work great on it is very very low.

but if you want you can post you computer spec (or laptop model) and we can search for any linux incompatibility.

**Comment by SirWindfield** [comment]
Just chiming in on the Linux distros. Ubuntu and PopOS are really straightforward to install and have a very similar user interface to Windows IMO. I wouldn't touch Arch if it's your very first Linux experience in over 20 years. It's hard to install, needs a lot of maintainence.

**Comment by erosman** [comment]
Thank you for the suggestions. [Best Linux Distributions That are Most Suitable for Beginners](https://itsfoss.com/best-linux-beginners/) also has similar suggestions.
TBH, I have to buy a new laptop, install Linux on it and gradually move to it from my current laptop.
Do Linux distros handle hardware drivers automatically?

**Comment by rezad1393** [comment]
> Thank you for the suggestions. [Best Linux Distributions That are Most Suitable for Beginners](https://itsfoss.com/best-linux-beginners/) also has similar suggestions.
> TBH, I have to buy a new laptop, install Linux on it and gradually move to it from my current laptop.
> Do Linux distros handle hardware drivers automatically?

this is just my opinion :
if you want to buy a new laptop that does not need heavy A.I. activity usage or something else that mandates nvidia usage on linux, then absolutly go for amd gpu or (a little bit weaker one) amd gpu.
I have a laptop from 10 years ago (yes it has beed that long but I upgraded the hdd to ssd and have 12 GB RAM) and it has intel gor cpu and gpu and nvidia for gpu as a dedicated card.
in the whole times I tried to use nvidia with linux, be it via optimus config and any other way, it was mostly a headache at best.
also my older pc got its gpu fried because nvidia released its updates manually and I forgot to update for a week and as my luck would have it I got the one driver with fan issue that fired the card.

also wayland (new display server on linux that is replacing x server) doesnt play nice at all with nvidia.
all told, because nvidia is a binary proriotery driver you WILL get issues but with amd if you get issues you can also count in open source part to fix it, and with intel it is even better support (intel doesnt have non-opensource gpu driver on linux at all).
so in my opinion basically avoid nvidia if you can.

other hardwares are usually supported unless it is something weird,niche or very very new.
like rgb leds or some tablet/laptop combos.
if the laptop you buy in not leading edge and very very new then you usually dont get errors or kernel panics in general.

if you want you can massage me with your desired laptop spec needs or a even better a specific laptop and I try to check it out on internet and sites that review them and give you back feedback before you commit to buy one.

**Comment by erosman** [comment]
> if you want to buy a new laptop that does not need heavy A.I. activity usage or something else that mandates nvidia usage on linux, then absolutly go for amd gpu or (a little bit weaker one) intel gpu.

Most of the ones I looked at were with Nvidia Geforce.

Upgrading from my current Acer Aspire V3-571G, i5 4GB RAM, I would like something i7, 16GB RAM, 512 SDD (no HDD), Gen10 (prefer Gen 11) CPU.

I have had a few Acer laptops and have been generally happy with them. I also had Dell but never again.
I am considering Acer or Lenovo :thinking:

**Comment by rezad1393** [comment]
> Most of the ones I looked at were with Nvidia Geforce.

I feel your pain.
I have the same issue too that I need to get a new laptop (this one is showing its age) and I absolutely dont want nvidia because at best it would just be a redundant unused part of laptop.
as I said, until nvidia open sources its driver I dont intend to buy from them.

but of course you can ignore that if you want but I would not recommend.

there are new laptop (intel 11 gen) with good integrated (not dedicated) gpu that would satisfy me but I live in iran so I have to wait a bit for them to get to market here.

the one I saw some months past was this :
https://www.notebookcheck.net/Lenovo-Yoga-Slim-7i-Carbon-13ITL5-laptop-review-Tiger-Lake-combined-with-ultralight-materials.530239.0.html#toc-verdict-the-lenovo-yoga-slim-7i-carbon-is-a-winner

btw this website is a good website that reviews laptops in house and in very detailed manner.

**Comment by SirWindfield** [comment]
You can still go NVidia if you don't care about opensource drivers. Their propriatary drivers are (as far as I remember) working fairly good. I got my 1070 working with them on Arch fairly easily.

However, Linux community seems to prever AMD GPUs due to them having fully opensource drivers.

**Comment by erosman** [comment]
I was also reading [4 Best Graphics Cards for Linux Users](https://www.addictivetips.com/buying-guides/best-graphics-card-for-linux/) which states:

> For this reason, many in the Linux community prefer to use AMD Radeon GPUs. Their drivers are open-source, included in the Linux kernel, and require no fiddling to get things working.

Lenovo Yoga is 13 inh. I would get a 15 inch laptop (I have been using 2 monitor for the last 25+ years, laptop + external).

I don't play games any more. I am not very worried about the budget and I was checking Lenovo Legion 5 :thinking:

**Comment by rezad1393** [comment]
> I was also reading [4 Best Graphics Cards for Linux Users](https://www.addictivetips.com/buying-guides/best-graphics-card-for-linux/) which states:

these are desktop cards but the driver opensource still applies.

not to confuse you but this is an issue like this:
https://old.reddit.com/r/linuxhardware/comments/go5eys/lenovo_thinkpad_e15_question/

but I think you are better served with a intel 11 gen instead of 10 gen.

**Comment by erosman** [comment]
AFA laptop support.... 
[Now You Can Buy Linux Certified Lenovo ThinkPad and ThinkStation (For the Best Possible Out of the Box Linux Experience)](https://itsfoss.com/lenovo-linux-certified/)

[Lenovo Brings Linux® Certification to ThinkPad and ThinkStation Workstation Portfolio, Easing Deployment for Developers & Data Scientists](https://news.lenovo.com/pressroom/press-releases/lenovo-brings-linux-certification-to-thinkpad-and-thinkstation-workstation-portfolio-easing-deployment-for-developers-data-scientists/)

**Comment by rezad1393** [comment]
this is old news but I didnt know you knew that much.
yes lenovo has parthnered with redhat (owner of fedora) to basically create linux certified laptops.
which is big news because the certification means backing and that allows companies it people to convince the higher ups to allow linux on devices of company.

this is a (I think) more that 6 month old news and the good part is that they are gonna do it retroactively for some lines on lenovo. I think thinkpads.

but I didnt pry much to that even though it is a good thing, because thinkpads are too expensive for me relative to other laptops (like acer or asus or huawei). but they are a good choice for corporations. also most thinkpads are not thin and light (though that is not always true).

this is one if the news ones with review (i7 and gen 11) 
https://www.notebookcheck.net/Lenovo-ThinkBook-15-G2-ITL-i7-1165G7.534044.0.html

**Comment by erosman** [comment]
If you have any suggestions for a top spec laptop suitable for Ubuntu installation, please let me know.

**Comment by rezad1393** [comment]
> If you have any suggestions for a high spec laptop suitable for Ubuntu installation, please let me know.

so ram 16 and higher cpu 11 gen intel or a good amd .
but what about display size and weight and number of ports (do you need thunderbolt or not).

**Comment by erosman** [comment]
> so ram 16 and higher cpu 11 gen intel or a good amd .
but what about display size and weight and number of ports (do you need thunderbolt or not).

I don't have anything that uses thunderbolt.
Since I always use a second monitor, a 14inch would also work but almost all 14" laptops are low specs. Therefore 15.6" would be best.

**Comment by rezad1393** [comment]
and you are against dell right?

**Comment by erosman** [comment]
I haven had a good experience with Dell.
I prefer lighter/slimmer ones if possible but specs are important.

**Comment by rezad1393** [comment]
https://www.laptoparena.net/laptop-finder?Internal_memory=16-128_GB&Display_diagonal=15-21_inch&Total_storage_capacity=500-5120_GB&Brand=Acer,Lenovo&Processor_family=11th_gen_Intel_Core_i7&Discrete_graphics_adapter=N

**Comment by rezad1393** [comment]
https://www.notebookcheck.net/Lenovo-ThinkPad-L15-laptop-review-Decent-office-companion-with-high-system-performance.503342.0.html

if you can wait some time,at most 3 months, then the new amd and intel cpus come and then the older ones (last year one) get more available.
https://www.crn.com/slide-shows/components-peripherals/9-business-laptops-with-the-new-amd-ryzen-5000-cpus/3

**Comment by rezad1393** [comment]
if you have a matrix username we can chat over there.
this is not really suitable for laptop hunting.

**Comment by erosman** [comment]
Is **Intel Iris Xe Graphics** supported by Ubuntu?

**Comment by rezad1393** [comment]
> Is **Intel Iris Xe Graphics** supported by Ubuntu?
absolutely
> I dont have matrix username.
it is a great network
even mozilla replaced irc with matrix.
https://wiki.mozilla.org/Matrix

**Comment by erosman** [comment]
I am on https://chat.mozilla.org/ (same username)

**Comment by rezad1393** [comment]
chatted you there

**Comment by afontenot** [comment]
Hey @erosman I don't have time to read every comment in this thread so I don't know exactly where this stands, but I'm excited to see that there's a userscript manager that uses the userScripts API in Firefox - and even better one that is actively receiving new development time. Thanks for your work, I was actually thinking about creating one myself until I found this.

> Greasemonkey on Firefox (283,000+ users) code is clear and on github but it is worked on by more than one person

I think this has the possibility of being a self-fulfilling prophecy. If there's a barrier to entry for contributions (on Github, sending your changes to be merged requires one click) then you can probably count on not receiving many contributions. Code that only ever gets uploaded to AMO feels like a private project, where contributions from others will not be welcomed, even if that's untrue.

If you're not comfortable using git or tracking every change doesn't fit well with your workflow, can I suggest the following approach as a stopgap measure? Create a repository with the code for Firemonkey (good name by the way), and only update it whenever you cut a new release on AMO. You can just stick every change in one commit for now. If you make this change, at least there will be code representing all the AMO releases on Github.

This would have the additional advantage of making it easier to find problems introduced by new releases.

Thanks for your time!

**Comment by sealor** [comment]
After reading all the comments I noticed that many comments already pointed out why it would be very useful to have the source code here on Github or any other code publishing platform. I also understood that using Github would hinder the workflow of the author.

So, we need an idea which fulfils both interests.

Does anybody know how Github actions work?
Is it possible to regularly download, extract and commit the XPI into a git repository?

**Comment by erosman** [comment]
TBH, I have thought about uploading some (3-4) core files of FireMonkey so that they can be viewed.

There is no point in uploading everything e.g. CodeMirror, libs etc. However, that would have lead to loads of complaint from people who like to clone (fork) the code as their own repository, since those files alone wouldn't be enough to be packed into a compete add-on.

I have already seen many examples of people who fork, rename and release the identical code as their own. There have been tens of thousands of such instances on AMO (and they get removed by AMO).

I surmised that those who are programmers, know how to inspect the code from the XPI, and those who are not programmers wouldn't need/want to inspect the code anyway.

**Comment by Iey4iej3** [comment]
I think it better to start with seeing whether versioning control tools like `git` help, which is completely local (nothing to upload). If it does help, then it might be worthwhile to "upload" to GitHub.

**Comment by bb010g** [comment]
if you'd be interested in using a GUI Git porcelain over a command-line porcelain, [Sublime Merge](https://www.sublimemerge.com/) is excellent.

The main reason I'd be interested in having version-tracked source easily readable online is for easily tracking how things have changed over time, without probably creating my own Git repository out of a bunch of extracted XPIs, and for being able to easily submit patches if I find a problem. (You wouldn't be obligated to accept the patches, but I personally like to try to fix a thing if it's broken and the source is available to me, to make maintainers' jobs a bit easier. If anything, it's a bit more documentation about my problem and what a partial fix might be.)

I'd also be willing to help you set up a historical Git repository, if that's something you'd be interested in.

**Comment by erosman** [comment]
As a trial ... I created a branch and release... 
Note: contains only main FM files

https://github.com/erosman/support/tree/FireMonkey

**Comment by Iey4iej3** [comment]
> As a trial ... I created a branch and release... Note: contains only main FM files
> 
> https://github.com/erosman/support/tree/FireMonkey

It seems that you "add files by upload", but it should be as simple as `git commit -am "XXX"` and `git push`.

**Comment by erosman** [comment]
I have uploaded the code as a trial, for discussion only. It is not a source code for reproduction. I do NOT develop on Github and I often made 100s of changes in a single day. It is not feasible for me to to reflect them here.
I do not use command line and I am currently on windows.
This repo is SUPPORT for all my addons.
FireMonkey is a branch of that support.

**Comment by pintassilgo** [comment]
If it's useful to someone:

[diff v2.37](https://github.com/pintassilgo/forked-firefox-extensions/commit/7b624d37f5fe3c913eedcaf1d0d860450dd95f3a)

[diff v2.36](https://github.com/pintassilgo/forked-firefox-extensions/commit/271e9675995035077a5c3c35dca9eae792913d5e)

[diff v2.35](https://github.com/pintassilgo/forked-firefox-extensions/commit/7ae2c1cfeb12fade451178f33196ea8df82d9821)

...

**Comment by bb010g** [comment]
@erosman Thank you! Adding files by upload on each release is fine, if that works for you.

> It seems that you "add files by upload", but it should be as simple as `git commit -am "XXX"` and `git push`. And seemingly, it should be a repo instead of a branch.

@Iey4iej3 A branch works fine. Git itself doesn't really care. You can have completely unrelated branches in the same repository; this is how GitHub Pages branches like to work.

**Comment by Iey4iej3** [comment]
> @Iey4iej3 A branch works fine. Git itself doesn't really care. You can have completely unrelated branches in the same repository; this is how GitHub Pages branches like to work.

It works, of course, but this does not seem to be what it was designed to be — different branches share most of files, and one could switch back and forth between different branches to do testings. Let's stop here — the developer does not seem to be interesting in using `git` (or any CVS system) as their developing tool.

**Comment by hshpm** [comment]
Just checked this out as it seems to be a great project. 
Worked like a charm so far. Then i saw a typo and wanted to do a quick pull request.  
Couldn't find the source code anywhere but in the xpi I got it from... Found this thread and in a couple of minutes reading through this my opinion of this project went down fast. All those useless arguments against collaboration and potential code-stealing is exactly the talk you hear from the companies you try to avoid coming here.

Just to be clear: You got nothing to lose, but everything to gain. Every idiot can unzip your extension and steal, but coders actually want to contribute. For stealing there are licenses to protect your property. For making a product greatt there is community contribution. Good software spreads being recommended by experts. Experts coming here will most likely not recommend a project refusing to use git.

TLDR:
Wanted to do a pull request, saw only a few files are checked in, went away sad, continue to use alternative software.

**Comment by erosman** [comment]
> Just to be clear: You got nothing to lose

Time.
Isn't time valuable?

> Every idiot can unzip your extension 

True.
Therefore Github source code is not imperative. 

> Wanted to do a pull request

The main files are already on github https://github.com/erosman/support/tree/FireMonkey

**Comment by adamency** [comment]
I made a git repository of all the history of the project I could find: https://github.com/adamency/firemonkey

I explained my reasoning and motive in detail here: https://github.com/erosman/support/issues/415#issuecomment-1865520465
